content/en/blog/_posts/2024-11-11-dynamic-kubernetes-api-server-for-cozystack/index.md

```diff
@@ -0,0 +1,454 @@
+  (with CRDs, you'd have to write webhooks for extended validation).
+  
+  You can implement transparent conversion logic between versions like `v1alpha1`, `v1beta1`, and `v1`
+  and serve multiple API versions simultaneously.
+  
+  
+  {{% alert title="Note" color="info" %}}
+  Kubernetes allows you to serve multiple API versions for each resource type, traditionally
+  `v1alpha1`, `v1beta1` and `v1`. Only one version can be specified as the storage version.
+  All requests to other versions must be automatically converted to the version specified as storage version.
+  With CRDs, this mechanism is implemented using conversion webhooks. Whereas in extension api-server,
+  this logic can be implemented directly.
+  {{% /alert %}}
+
+- The api-server lets you format table output however you like and doesn't force you to follow
+  the `additionalPrinterColumns` logic in CRDs. Instead, you can write your own formatter that
+  formats the table output and custom fields in it. For example, when using `additionalPrinterColumns`,
+  you can display field values only following the JSONPath logic. In your own api-server, you can generate
+  and insert values on the fly, formatting the table output as you wish.
+
+### Dynamic resource registration
+
+- The resources served by an extension api-server don't need to be pre-registered as CRDs.
+Once your api-server is registered as an `APIService`, Kubernetes starts polling it to find out what
+resource types it serves. After receiving a response, it automatically registers all available
+types in its API. Consequently, you can implement logic that automatically registers the resource
+types you need in Kubernetes.
+
+## When not to use the API Aggregation Layer
+
+There are some anti-patterns where using the API Aggregation Layer isn't recommended.
+Let's go through them.
+
+### Unstable backend
+
+If your api-server stops responding for some reason due to an unavailable backend or other issues it
+may block some Kubernetes functionality. For example, when deleting namespaces, Kubernetes will wait
+for a response from your api-server to see if there are any remaining resources.
+If the response doesn't come, the namespace deletion will be blocked.
+
+Also, you might have encountered a [situation](https://github.com/kedacore/keda/issues/4224) where,
+when the metrics-server is unavailable, an extra message appears in stderr after every API request
+(even unrelated to metrics) stating that `metrics.k8s.io` is unavailable. This is another example
+of how using the API Aggregation Layer can lead to problems when the api-server handling requests
+is unavailable.
+
+### Slow requests
+
+If you can't guarantee an instant response for user requests, it's better to consider using the operator
+pattern. Otherwise, you might make your cluster less stable. Many projects implement an extension
+api-server only for a limited set of resources, particularly for imperative logic and subresources.
```

Comment:
```suggestion
If you can't guarantee an instant response for user requests, it's better to consider using a
CustomResourceDefinition and controller.
Otherwise, you might make your cluster less stable. Many projects implement an extension
API server only for a limited set of resources, particularly for imperative logic and subresources.
```